subtests/docker_cli/systemd/systemd_run.py

```diff
@@ -0,0 +1,77 @@
+"""
+Summary
+-------
+Testing basic container as systemd service
+
+Operational Summary
+---------------------
+This is to test container as systemd service using a predefined unit file
+(``p4321.service``) and a ``script p4321-server.py``. An image is built with
+this script added. Systemd starts a container using this image & the container
+writes current time to port ``4321``. From host, the socket is read and the
+value is checked to see if it is a digit. Finally, container is stopped
+through systemd and system is cleaned up.
+
+"""
+
+import time
+import socket
+from os.path import join
+from shutil import copy
+from shutil import copyfile
+from autotest.client import utils
+from autotest.client.shared.utils import is_port_free
+from dockertest.dockercmd import DockerCmd
+from dockertest.output import mustpass
+from dockertest.xceptions import DockerTestError
+from systemd import systemd_base
+
+
+class systemd_run(systemd_base):
+
+    def initialize(self):
+        # edit unitfile & copy it to /etc/systemd/system/
+        super(systemd_run, self).initialize()
+        self.write_unitfile(self.sub_stuff['vars_dict'])
+        copyfile(self.sub_stuff['tmpdirf'], self.sub_stuff['unitf_dst'])
+        # copy Dockerfile & p4321-server.py to tmpdir and edit Dockerfile
+        tmpdir = self.sub_stuff['tmpdir']
+        self.sub_stuff['tmpdirf'] = join(tmpdir, 'Dockerfile')
+        for f in 'Dockerfile', 'p4321-server.py':
+            copy(join(self.sub_stuff['bindir'], f), tmpdir)
+        self.sub_stuff['vars_dict'] = {'FROM ': self.config['img_frm']}
+        self.write_unitfile(self.sub_stuff['vars_dict'])
+        # build image using edited Dockerfile in tmpdir
+        dkrcmd = DockerCmd(self, 'build', [self.config['build_opt'], tmpdir])
+        mustpass(dkrcmd.execute())
+
+    def run_once(self):
+        host = 'localhost'
+        port = 4321
+        self.sysd_action('daemon-reload')
+        self.sysd_action('start', self.config['sysd_unit_file'])
+        utils.wait_for(lambda: not is_port_free(port, host), 10)
+        time_from_socket = self.read_socket(host, port)
+        self.failif(not time_from_socket.isdigit(),
+                    "Data received from container is non-numeric: '%s'"
+                    % time_from_socket)
+
+    @staticmethod
+    def read_socket(host, port):
+        max_read_tries = 10
+        num_read_try = max_read_tries
+        while num_read_try:
+            # using timeout 13-sec as it's 1-sec longer than two DNS timeouts
+            s = socket.create_connection((host, port), 13)
+            s.settimeout(2.0)
+            try:
+                time_from_socket = s.recv(512).strip()
+                if time_from_socket:
+                    s.shutdown(socket.SHUT_RDWR)
+                    return time_from_socket
+                num_read_try -= 1
+            except socket.timeout:
+                pass
+            time.sleep(1.0)
```

Comment:
@cevich when I do that, I am getting "Failed to read from socket with %d tries" error frequently. So how about:
```
            finally:
                time.sleep(1.0)
            skt.shutdown(socket.SHUT_RDWR)